modules/proto/src/smithytranslate/proto3/internals/Compiler.scala

```diff
@@ -22,21 +22,61 @@ import software.amazon.smithy.model.Model
 import software.amazon.smithy.model.shapes._
 import software.amazon.smithy.model.traits.DeprecatedTrait
 import software.amazon.smithy.model.traits.EnumTrait
-import software.amazon.smithy.model.traits.RequiredTrait
-import software.amazon.smithy.model.traits.SparseTrait
 import software.amazon.smithy.model.traits.UnitTypeTrait
 import software.amazon.smithy.model.traits.TraitDefinition
 
 import scala.jdk.CollectionConverters._
 import scala.jdk.OptionConverters._
+import software.amazon.smithy.model.neighbor.NeighborProvider
+import software.amazon.smithy.model.neighbor.Walker
+import alloy.OpenEnumTrait
+import software.amazon.smithy.model.traits.EnumValueTrait
 
-class Compiler() {
+private[proto3] class Compiler(model: Model, allShapes: Boolean) {
 
   // Reference:
   // 1. https://github.com/protocolbuffers/protobuf/blob/master/docs/field_presence.md
 
   import ProtoIR._
 
+  private lazy val allRelevantShapes: Set[Shape] = {
+    if (allShapes) {
+      model
+        .shapes()
+        .iterator()
+        .asScala
+        .filterNot(ShapeFiltering.exclude)
+        .toSet
+    } else {
+      val walker = new Walker(NeighborProvider.of(model))
+      val protoEnabledShapes =
+        model.getShapesWithTrait(classOf[ProtoEnabledTrait]).asScala
+      val grpcShapes = model.getShapesWithTrait(classOf[GrpcTrait]).asScala
+      val allRoots = protoEnabledShapes ++ grpcShapes
+      val allTransitiveShapes: Set[Shape] = allRoots
+        .flatMap((shape: Shape) => walker.walkShapes(shape).asScala)
+        .toSet
+      (allRoots ++ allTransitiveShapes).toSet
```

Comment:
should we still be calling `.filterNot(ShapeFiltering.exclude)`

Reply:
resolved https://github.com/disneystreaming/smithy-translate/pull/232/files#diff-9090ee3dc5fe3560fc224d5149c9341a985ffec77ec48eab4afb679f494bc000R108